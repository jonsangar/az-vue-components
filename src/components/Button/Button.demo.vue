<template lang="pug">
  div
    Demo(title="Tipo")
      az-button Default
      az-button(type="brand") Brand
      az-button(type="success") Success
      az-button(type="warning") Warning
      az-button(type="danger") Danger

    Demo(title="Plano")
      az-button(plain) Default
      az-button(type="brand" plain) Brand
      az-button(type="success" plain) Success
      az-button(type="warning" plain) Warning
      az-button(type="danger" plain) Danger

    Demo(title="Ghost")
      az-button(ghost) Default
      az-button(type="brand" ghost) Brand
      az-button(type="success" ghost) Success
      az-button(type="warning" ghost) Warning
      az-button(type="danger" ghost) Danger

    Demo(title="Forma")
      az-button(type="brand") Default
      az-button(type="brand" circle) Circle
      az-button(type="brand" square) Square

    Demo(title="Disabled")
      az-button(type="brand" disabled) Disabled
      az-button(type="success" disabled) Disabled

    Demo(title="Block")
      az-button.no-margin.margin-bottom-half(type="brand" block) Block Brand
      az-button.no-margin(type="danger" block) Block Danger

    Demo(title="Tama√±o")
      az-button(type="brand" size="small") Small
      az-button(type="brand") Default
      az-button(type="brand" size="large") Large

    Demo(title="Icono")
      .demo__subtitle Icon and text
      az-button(type="brand" iconName="palm") Icon 1
      az-button(type="brand" iconName="lap") Icon 2
      az-button(type="brand" iconName="desktop") Icon 3

      .demo__subtitle.margin-top Only Icon
      az-button(type="brand" iconName="palm" size="small")
      az-button(type="brand" iconName="palm")
      az-button(type="brand" iconName="palm" size="large")

      .demo__subtitle.margin-top Icon Position
      .padding-x-half.padding-top
        az-row(grid)
          az-button(type="brand" iconName="palm") Default
          az-button(type="brand" iconName="palm" iconPosition="right") Right
        az-row(grid)
          az-button(type="brand" iconName="palm" iconPosition="top") Top
          az-button(type="brand" iconName="palm" iconPosition="bottom") Bottom

    Demo(title="Loading")
      az-button.margin-bottom(type="default" @click="toggleDemo") Toggle Loading

      .demo__subtitle Icon and text
      div
        az-button(type="brand" :loading="loading" loadingType="square") Guardar
      div
        az-button(type="brand" :loading="loading" loadingType="bars") Guardar
      div
        az-button(type="brand" :loading="loading" loadingType="circle") Guardar
      div
        az-button(type="brand" :loading="loading" loadingType="blobs") Guardar
      div
        az-button(type="brand" :loading="loading") Guardar

      .demo__subtitle Only Icon
      div
        az-button(type="brand" :loading="loading" loadingIconOnly) Guardar
      div
        az-button(type="brand" :loading="loading" loadingIconOnly loadingType="square") Guardar

      .demo__subtitle Change text
      div
        az-button(:loading="loading" loadingText="Buscando") Buscar
      div
        az-button(:loading="loading" loadingText="Configurando") Guardar

</template>

<script>
export default {
  data() {
    return {
      loading: false,
    };
  },
  methods: {
    toggleDemo() {
      this.loading = !this.loading;
    },
  },
};
</script>

<style lang="scss" scoped>
.az-button {
  margin: quarter();
}
</style>
