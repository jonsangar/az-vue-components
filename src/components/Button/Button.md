---
device: true
---

# Button

### Instalación

```javascript
import { Button } from 'didor';

Vue.use(Button);
```

---

### Uso

#### Tipo

```pug
az-button Default
az-button(type="brand") Brand
az-button(type="success") Success
az-button(type="warning") Warning
az-button(type="danger") Danger
```

#### Plain

```pug
az-button(plain) Default
az-button(type="brand" plain) Brand
az-button(type="success" plain) Success
az-button(type="warning" plain) Warning
az-button(type="danger" plain) Danger
```

#### Ghost

```pug
az-button(ghost) Default
az-button(type="brand" ghost) Brand
az-button(type="success" ghost) Success
az-button(type="warning" ghost) Warning
az-button(type="danger" ghost) Danger
```

#### Forma

```pug
az-button(type="brand") Default
az-button(type="brand" circle) Circle
az-button(type="brand" square) Square
```

#### Disabled

```pug
az-button(type="brand" disabled) Disabled
az-button(type="danger" disabled) Disabled
```

#### Block

```pug
az-button(type="brand" block) Block Brand
az-button(type="danger" block) Block Danger
```

#### Tamaño

```pug
az-button(type="brand" size="large") Large
az-button(type="brand") Default
az-button(type="brand" size="small") Small
```

#### Icono

```pug
az-button(type="brand" iconName="palm") Icon 1
az-button(type="brand" iconName="lap") Icon 2
az-button(type="brand" iconName="desktop") Icon 3

az-button(type="brand" iconName="palm")
az-button(type="brand" iconName="lap")
az-button(type="brand" iconName="desktop")

az-button(type="brand" iconName="palm") Default
az-button(type="brand" iconName="palm" iconPosition="top") Right
az-button(type="brand" iconName="palm" iconPosition="right") Top
az-button(type="brand" iconName="palm" iconPosition="bottom") Bottom
```

#### Loading

```pug
az-button(loading) Guardar
az-button(loading loadingType="square") Guardar
az-button(loading loadingType="bars") Guardar
az-button(loading loadingType="circle") Guardar

az-button(loading loadingText="Buscando") Buscar
az-button(loading loadingText="Configurando") Guardar

az-button(loading loadingIconOnly) Guardar
```

---

### API

| Attribute   | Description                                | Type      | Default   |
| ----------- | ------------------------------------------ | --------- | --------- |
| type        | Tipo de botón, opciones:<br>`default` `brand` `success` `warning` `danger` | `String`  | `default` |
| plain       | Determinar si es o no un botón plano       | `Boolean` | `false`   |
| ghost       | Determinar si es o no un botón ghost       | `Boolean` | `false`   |
| circle      | Determina si los lateras son circulares    | `Boolean` | `false`   |
| square      | Determina si las esquinas son cuadradas    | `Boolean` | `false`   |
| block       | Determina si es o no un elemento de bloque | `Boolean` | `false`   |
| disabled    | Deshabilita el botón                       | `Boolean` | `false`   |
| size        | Tamaño del botón, opciones:<br>`large` `normal` `small`       | `String`  | `normal`  |
| iconName    | Indica el nombre de un icono               | `String`  | --        |
| iconPosition | Indica la posición del icono respecto del texto, opciones:<br>`top` `right` `bottom` `left`           | `String`  | `left`        |
| loading     | Determina si su estado es loading o no     | `Boolean` | `false`   |
| loadingType | Determina el tipo de icono de loading, opciones:<br>`star` `circle` `success` `bars` `blobs` `square`      | `String`  | `star`   |
| loadingText | Determina el texto que se mostrará en el estado loading     | `String` | `loading`   |
| loadingIconOnly | Determina si solo se mostrará el icono de loading, sin texto. | `Boolean` | `false`   |
| autofocus   | Misma función que el atributo nativo **autofocus** | `Boolean` | `false`   |
| native-type | Misma función que el atributo nativo **type**, opciones:<br> `button` `submit` `reset`                      | `String`  | `button`      |

### Event

| Event | Description                                             | Arguments |
| ----- | ------------------------------------------------------- | --------- |
| click | Triggered when click button and not disabled or loading | -         |
