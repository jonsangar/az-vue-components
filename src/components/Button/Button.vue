<template lang="pug">
  button.az-button(
    @click="handleClick"
    :disabled="disabled || loading"
    :autofocus="autofocus"
    :type="nativeType"
    :class="classObject"
    :style="width ? `min-width: ${width}` : ''"
    @touchstart="emitFeedback()")

    //- Loading spinner
    az-loading(:type="loadingType" size=".8em" v-if="loading")
    az-icon(:name="iconName" :scale="iconScale" v-if="iconName && !loading")
    span(v-if="$slots.default && (!loading || loadingIconOnly)")
      slot
    span(v-if="loading && !loadingIconOnly") {{ loadingText }}
</template>

<script>

export default {
  name: 'AzButton',

  props: {
    autofocus: {
      type: Boolean,
      default: false,
    },
    block: {
      type: Boolean,
      default: false,
    },
    circle: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    ghost: {
      type: Boolean,
      default: false,
    },
    iconName: {
      type: String,
      default: '',
    },
    iconPosition: {
      type: String,
      default: '',
      validator: value => ['top', 'right', 'bottom', 'left', ''].indexOf(value) !== -1,
    },
    iconAnimation: {
      type: String,
      default: '',
      validator: value => ['rotate', ''].indexOf(value) !== -1
    },
    iconScale: {
      type: String,
      default: '1'
    },
    loading: {
      type: Boolean,
      default: false,
    },
    loadingType: {
      type: String,
      default: 'star',
    },
    loadingIconOnly: {
      type: Boolean,
      default: false,
    },
    loadingText: {
      type: String,
      default: 'Cargando',
    },
    plain: {
      type: Boolean,
      default: false,
    },
    nativeType: {
      type: String,
      default: 'button',
      validator: value => ['button', 'submit', 'reset'].indexOf(value) !== -1,
    },
    size: {
      type: String,
      default: '',
    },
    square: {
      type: Boolean,
      default: false,
    },
    type: {
      type: String,
      default: '',
    },
    width: {
      type: String,
      default: ''
    }
  },
  computed: {
    classObject() {
      return [
        this.type ? `az-button--${this.type}` : '',
        this.size ? `az-button--${this.size}` : '',
        this.iconName && this.iconPosition ? `az-button--icon-${this.iconPosition}` : '',
        this.iconName && this.iconAnimation ? `az-button--icon-${this.iconAnimation}` : '',
        {
          'az-button--icon': this.iconName,
          'az-button--icon-only': !this.$slots.default,
          'az-button--disabled': this.disabled,
          'az-button--loading': this.loading,
          'az-button--loadingIconOnly': this.loadingIconOnly,
          'az-button--block': this.block,
          'az-button--ghost': this.ghost,
          'az-button--plain': this.plain,
          'az-button--square': this.square,
          'az-button--circle': this.circle,
        },
      ];
    },
  },
  methods: {

    handleClick(event) {
      this.$emit('click', event);
    },
  },
};
</script>
