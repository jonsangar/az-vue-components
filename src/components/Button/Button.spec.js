import { mount, createLocalVue } from '@vue/test-utils';
// Importamos librerías de estilos e iconos
import SvgIcon from 'vue-svgicon'
import AzLoading from 'Core/components/Loading/Loading.vue';
import Button from './Button.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

localVue.component('AzLoading', AzLoading);

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente Button con snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toMatchSnapshot()

  })

  it('Comprueba que emite evento "click" al ser pulsado', async () => {
    const { wrapper } = build();
    // Al pulsar sobre el elemento lanza el evento click
    wrapper.trigger('click');
    // Comprobamos que se ha emitido el evento click
    expect(wrapper.emitted().click).toBeTruthy();
    expect(wrapper.emitted().click.length).toBe(1);
  });

  it('Comprueba que el botón tiene la propiedad disabled activada', async () => {
    const { wrapper } = build();
    // Si actualizamos la propiedad disabled, añade la clase disabled
    await wrapper.setProps({
      disabled: true,
    });

    // Comprobamos que se ha desabilitado la prop
    expect(wrapper.vm.disabled).toStrictEqual(true);
    // Comprobamos que se ha añadido la clase desabilitado
    expect(wrapper.find('.az-button--disabled').isVisible()).toBe(true);
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        loading: true,
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad loading activada', async () => {
    const { wrapper } = build();
    // Comprobamos que se ha activado la prop
    expect(wrapper.vm.loading).toStrictEqual(true);
    // Comprobamos que se ha añadido la clase loading
    expect(wrapper.find('.az-button--loading').isVisible()).toBe(true);
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        loadingIconOnly: true
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad loadingIconOnly activada', async () => {
    const { wrapper } = build();
    // Comprobamos que se ha activado la prop
    expect(wrapper.vm.loadingIconOnly).toStrictEqual(true);
    // Comprobamos que se ha añadido la clase loadingIconOnly
    expect(wrapper.find('.az-button--loadingIconOnly').isVisible()).toBe(true);
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        block: true,
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad block activada', async () => {
    const { wrapper } = build();
    // Comprobamos que se ha activado la prop
    expect(wrapper.vm.block).toStrictEqual(true);
    // Comprobamos que se ha añadido la clase block
    expect(wrapper.find('.az-button--block').isVisible()).toBe(true);
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        ghost: true,
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad ghost activada', async () => {
    const { wrapper } = build();
    // Comprobamos que se ha activado la prop
    expect(wrapper.vm.ghost).toStrictEqual(true);
    // Comprobamos que se ha añadido la clase ghost
    expect(wrapper.find('.az-button--ghost').isVisible()).toBe(true);
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        plain: true,
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad plain activada', async () => {
    const { wrapper } = build();
    // Comprobamos que se ha activado la prop
    expect(wrapper.vm.plain).toStrictEqual(true);
    // Comprobamos que se ha añadido la clase ghost
    expect(wrapper.find('.az-button--plain').isVisible()).toBe(true);
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        square: true,
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad square activada', async () => {
    const { wrapper } = build();
    expect(wrapper.vm.square).toStrictEqual(true);
    // Comprobamos que se ha añadido la clase square
    expect(wrapper.find('.az-button--square').isVisible()).toBe(true);
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        circle: true,
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad circle activada', async () => {
    const { wrapper } = build();
    expect(wrapper.vm.circle).toStrictEqual(true);
    // Comprobamos que se ha añadido la clase circle
    expect(wrapper.find('.az-button--circle').isVisible()).toBe(true);
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        type: 'rafa'
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad type activada', async () => {
    const { wrapper } = build();
    // Comprobamos que se ha activado la prop
    expect(wrapper.vm.type).toStrictEqual('rafa');
    // Comprobamos que se ha añadido la clase circle
    expect(wrapper.find('.az-button--rafa').isVisible()).toBe(true);
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        size: 'sm'
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad size activada', async () => {
    const { wrapper } = build();
    // Comprobamos que se ha activado la prop
    expect(wrapper.vm.size).toStrictEqual('sm');
    // Comprobamos que se ha añadido la clase circle
    expect(wrapper.find('.az-button--sm').isVisible()).toBe(true);
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        iconName: 'success',
        iconPosition: 'left'
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad iconName e iconPosition', async () => {
    const { wrapper } = build();
      // Comprobamos que se ha activado la prop
    expect(wrapper.vm.iconName).toStrictEqual('success');
    expect(wrapper.vm.iconPosition).toStrictEqual('left');
    // Comprobamos que se ha añadido la clase circle
    expect(wrapper.find('.az-button--icon-left').isVisible()).toBe(true);
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        nativeType: 'submit'
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad nativeType', async () => {
    const { wrapper } = build();
    // Comprobamos que se ha activado la prop
    expect(wrapper.vm.nativeType).toStrictEqual('submit');
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        loadingText: 'montes'
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad loadingText', async () => {
    const { wrapper } = build();
    // Comprobamos que se ha activado la prop
    expect(wrapper.vm.loadingText).toStrictEqual('montes');
  });
})

describe('Button.vue', () => {
  const build = () => {
    const wrapper = mount(Button, {
      localVue,
      propsData: {
        loadingType: 'myLoading'
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que el botón tiene la propiedad loadingType', async () => {
    const { wrapper } = build();
    // Comprobamos que se ha activado la prop
    expect(wrapper.vm.loadingType).toStrictEqual('myLoading');
  });
})
