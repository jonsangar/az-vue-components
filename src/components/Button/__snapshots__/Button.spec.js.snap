// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue Comprueba renderizado del componente Button con snapshot 1`] = `
<button type="button" class="az-button az-button--icon-only">
  <!---->
  <!---->
  <!---->
  <!---->
</button>
`;
